swap/peer.go

```diff
@@ -137,3 +120,94 @@ func (sp *Peer) handleErrorMsg(ctx context.Context, msg interface{}) error {
 	// maybe balance disagreement
 	return nil
 }
+
+// processAndVerifyCheque verifies the cheque and compares it with the last received cheque
+// if the cheque is valid it will also be saved as the new last cheque
+func (sp *Peer) processAndVerifyCheque(cheque *Cheque) error {
+	if err := sp.verifyChequeProperties(cheque); err != nil {
+		return err
+	}
+
+	lastCheque := sp.loadLastReceivedCheque()
+
+	// TODO: there should probably be a lock here?
+	expectedAmount, err := sp.swap.oracle.GetPrice(cheque.Honey)
+	if err != nil {
+		return err
+	}
+
+	if err := sp.verifyChequeAgainstLast(cheque, lastCheque, expectedAmount); err != nil {
+		return err
+	}
+
+	if err := sp.saveLastReceivedCheque(cheque); err != nil {
+		log.Error("error while saving last received cheque", "peer", sp.ID().String(), "err", err.Error())
+		// TODO: what do we do here?
```

Comment:
One option would be to retry several times. I could think of a queue which tries to save it, while at least save it in memory (here the map actually becomes partially useful...)

Reply:
What do we do with the cached value in the meantime? Right now even if the saving fails the cached value is still updated.

If we retry several times what should happen if we get a new cheque but the old one hasn't even been saved yet?

Reply:
Let's not discuss this here. This issue has a lot of ramifications and possible solutions, there is a general issue for handling error situations: https://github.com/ethersphere/swarm/issues/1601

We can handle this there, or, if you prefer a dedicated issue just for this case so we don't forget, feel free to create a new issue.

Reply:
So for this PR should we ignore the error like now (and only update memory) or return the error without updating balances?

Reply:
Actually this is related to this issue: https://github.com/ethersphere/swarm/issues/1515

If we can't save the cheque, then there is a serious problem with the file system or something, and we should shut down the node. In that case the update to memory would be lost anyway too. So for this PR we can ignore the error, but we need to address this when we do https://github.com/ethersphere/swarm/issues/1515

Reply:
What about updating memory only if the `Put` method is successful?

Reply:
It's a possibility but then the next received cheque (and any that follows) will be rejected as the increase won't match the honey amount.